src/NUnitFramework/framework/Internal/Execution/WorkShift.cs

```diff
@@ -161,11 +161,11 @@ private void StartWorkers()
                 worker.Idle += (s, ea) =>
                 {
                     // Quick check first using Interlocked.Decrement
-                    if (Interlocked.Decrement(ref _busyCount) == 0)
+                    if (Interlocked.Decrement(ref _busyCount) == 0 && !HasWork)
```

Comment:
Two concerns with `HasWork` outside a `lock`: 
 - In the worst case scenario, the current thread's view of the list memory behind `HasWork` is stale and never updates to become `false`, so the `if` branch is never executed. We need a memory barrier which brings me to the next point:
 - Reading the `IList` can result in corrupted results (dirty read, tearing- I've observed the weird internal exceptions caused by this in practice) if the `IList` is being changed by another thread. Either we need an absolute guarantee that the `IList` itself is a thread safe collection (which solves the first point as well), or we need to lock read access as well as write access to it (which again solves the first point).

Reply:
And I see it's backed by a `List<>` which means we definitely can't read it without locking access.

Reply:
This could be a source of some intermittent issues I'm seeing. I'll check.

BTW, this is far from done!

Reply:
@jnm2 Thinking about this, I realized that I'm relying on the fact that the underlying list is never changed once set up. A lot of the parallel dispatcher code is like that...

* `ParallelWorkItemDispatcher` has fixed lists of shifts, queues and workers.
* `WorkShift` has a specific set of queues and workers.
* `WorkItemQueue` has a specific set of workers.

For that reason, I've played a little loose with locking. Maybe it would be better to use a collection that can't be updated at all, just in case.

Reply:
Ah, and I'm so used to looking for `IReadOnlyList<>` to make an appearance for that guarantee... It would simplify my understanding of the code if there was a builder type. How much code would break if `WorkShift` stopped exposing the `IList<>` and made a defensive array copy of the list passed into the constructor, stored as a private readonly `WorkItemQueue[]` and an XML comment on it that it should be immutable? (Or we could even add a lightweight `ImmutableArray` to our BCL-like toolbox?)

So long as no thread mutates the `IList<>` after the first time another thread reads it, we don't need to lock access to it.

Reply:
Queues is only public for test purposes. Could easily be internal. I'll take a look at the other collections of queues, shifts and workers as well. This may be an easy fix.

Reply:
For now, I'm only changing this to internal. In fact, determining whether a shift is really over is the key to a lot of our issues so I expect to be back in this bit of code in subsequent commits as I add tests for the various issues.

Reply:
I got sidetracked talking about the list, but we still have a staleness issue checking `WorkItemQueue.IsEmpty`. We need locked access to that mutable state.

Reply:
[Just checked on it.](https://github.com/nunit/nunit/blob/0162e7c1d8640769be2714ea81a3002903f36617/src/NUnitFramework/framework/Internal/Execution/WorkItemQueue.cs#L169-L179) No staleness issue because you're looking through an array (which I assume is also mot mutated once multiple threads see it) and each array item is a ConcurrentQueue.

Though on a new topic now, having more than one concurrency primitive in action makes me feel like everything is susceptible to race conditions. What happens if one of the queues gets an item added on a background thread right after `IsEmpty` returns false?